### PR TITLE
ULMFIT with fastai-v1

### PR DESCRIPTION
**Pull request by maaario**
now only to show you my progress on data processing and model pre-training

### CODE REVIEW

**Comment by buj** [comment]
### main.py

`prepare_lm_dataset`: I believe it either does not tokenize the input at all, or it uses some default tokenizer (i.e. for english language). So when creating the TextLMDataBunch we should pass it `tokenizer=Tokenizer(lang = "xx")` or something like that (`xx` stands for any obscure language, and slovak language is an obscure language).

`prepare_clas_dataset`: Same as above.

Minor note: one need not pass `test_df=None`, it is `None` by default.

### util.py

`evaluate_perplexity`: I highly suspect that it is possible to calculate perplexity simply by evaluating the model on the dataset with an appropriate loss function. See [this link](https://en.wikipedia.org/wiki/Perplexity#Perplexity_of_a_probability_model)---it is merely `e**cross_entropy`, so it should be easy to calculate using cross-entropy loss. The loss can be obtained using [`Learner.validate(...)`](https://docs.fast.ai/basic_train.html#Learner.validate).

### Other remarks

Currently the learner does not use CUDA (?). The model will sit on the same device as the dataset's (or rather databunch's) device, which is currently CPU (there is no flag to set it otherwise). There are two ways to alter this behaviour:
- Work with [`DeviceDataLoader`](https://docs.fast.ai/basic_data.html#DeviceDataLoader)s, but this is a big step away from what is currently implemented. Currently we load the data directly through pandas; to use the DeviceDataLoader, one must first construct a Pytorch [`DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader).
- When constructing the databunch from pandas dataframe, one can pass the argument `device = (some torch device)`. (This is not obvious from the documentation, but should work---`from_df` internally uses `DataBunch.create` and it passes keyword arguments to that constructor.)

File: ulmfit-fastai-v1/main.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+python main.py train_lm lm_dataset trained_lm --epochs=2
+
+python main.py lm_perplexity zlocin-a-trest trained_lm
+python main.py lm_perplexity lm_dataset trained_lm
+
+python main.py prepare_clas_dataset clas-alza.csv clas_dataset
+"""
+
+
+def prepare_lm_dataset(input_path, output_dir, valid_split=0.2, min_freq=2):
```
Comment (buj) [inline ulmfit-fastai-v1/main.py]: I believe it either does not tokenize the input at all, or it uses some default tokenizer (i.e. for english language). So when creating the `TextLMDataBunch` we should pass it `tokenizer=Tokenizer(lang = "xx")` or something like that (`xx` stands for any obscure language, and slovak language is an obscure language).

File: ulmfit-fastai-v1/main.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        output_dir, train_df, valid_df, test_df=None, text_cols=0, min_freq=min_freq)
+    data_lm.save('data_lm.pkl')
+
+    with open(output_dir / "data_lm_tokenized_train.txt", "w") as f:
+        f.write("\n".join(map(str, list(data_lm.train_ds.x))))
+    with open(output_dir / "data_lm_tokenized_valid.txt", "w") as f:
+        f.write("\n".join(map(str, list(data_lm.valid_ds.x))))
+
+
+def prepare_clas_dataset(input_path, output_dir, valid_split=0.2):
```
Comment (buj) [inline ulmfit-fastai-v1/main.py]: Same as above.

File: ulmfit-fastai-v1/main.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    """
+    train_df, valid_df = csv_to_train_valid_df(input_path, valid_split)
+
+    if not os.path.exists(output_dir):
+        os.mkdir(output_dir)
+    output_dir = Path(output_dir)
+
+    data_finetune_lm = TextLMDataBunch.from_df(output_dir, train_df, valid_df, test_df=None)
+    data_clas = TextClasDataBunch.from_df(
+        output_dir, train_df, valid_df, test_df=None, vocab=data_finetune_lm.train_ds.vocab, bs=32)
```
Comment (buj) [inline ulmfit-fastai-v1/main.py]: Minor note: one need not pass `test_df=None`, it is `None` by default.

File: ulmfit-fastai-v1/utils.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    df = pd.read_csv(csv_path, header=None)
+
+    df = df.iloc[np.random.permutation(len(df))]
+    cut = int(valid_split * len(df)) + 1
+    train_df, valid_df = df[cut:], df[:cut]
+
+    return train_df, valid_df
+
+
+def evaluate_perplexity(learner, text):
```
Comment (buj) [inline ulmfit-fastai-v1/utils.py]: I highly suspect that it is possible to calculate perplexity simply by evaluating the model on the dataset with an appropriate loss function. See this link---it is merely `e**cross_entropy`, so it should be easy to calculate using cross-entropy loss. The loss can be obtained using [`Learner.validate(...)`](https://docs.fast.ai/basic_train.html#Learner.validate).

**Review by buj — COMMENTED** [review]
### Other remarks

Currently the learner does not use CUDA (?). The model will sit on the same device as the dataset's (or rather databunch's) device, which is currently CPU (there is no flag to set it otherwise). There are two ways to alter this behaviour:
- Work with [`DeviceDataLoaders`](https://docs.fast.ai/basic_data.html#DeviceDataLoader), but this is a big step away from what is currently implemented. Currently we load the data directly through pandas; to use the DeviceDataLoader, one must first construct a Pytorch [`DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader).
- When constructing the databunch from pandas dataframe, one can pass the argument device = (some torch device). (This is not obvious from the documentation, but should work---from_df internally uses [`DataBunch.create`](https://docs.fast.ai/basic_data.html#DataBunch.create) and it passes keyword arguments to that constructor.)

File: ulmfit-fastai-v1/main.py (hunk 1 above)
Comment (maaario) [reply ulmfit-fastai-v1/main.py]: Thanks. By defualt english tokenizer is used, but we can switch to xx as you described. This will make tokenization more explicite and we can potentially tweak our own tokenizer in the future.

File: ulmfit-fastai-v1/main.py (hunk 2 above)
Comment (maaario) [reply ulmfit-fastai-v1/main.py]: fixed

File: ulmfit-fastai-v1/main.py (hunk 3 above)
Comment (maaario) [reply ulmfit-fastai-v1/main.py]: thanks, fixed

File: ulmfit-fastai-v1/utils.py (hunk 4 above)
Comment (maaario) [reply ulmfit-fastai-v1/utils.py]: Now we use e**loss as the main perplexity estimate, but I also kept our perplexity computation, because we can control it well.

File: ulmfit-fastai-v1/finetune_lm.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import sys, fire
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: split to separate imports - fire to 2nd group (3rd party)

File: ulmfit-fastai-v1/finetune_lm.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+from fastai.text import *
+from utils import *
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: put utils to separate import group as it is our local library

File: ulmfit-fastai-v1/finetune_lm.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def finetune_lm(data_dir, model_dir, cyc_len = 25, lr = 4e-3, lr_factor = 1/2.6):
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: PEP8 - no spaces around "=" in function headers, add spaces around operator "/"

File: ulmfit-fastai-v1/finetune_lm.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    """
+    Finetunes the provided language model on the given data, uses
+    techniques from ULMFIT paper (Discriminative fine-tuning, Slanted
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: As our whole project is based on ULMFIT, this is unnecessary here.

File: ulmfit-fastai-v1/finetune_lm.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    triangular learning rates).
+    
+    <data_dir>: the directory from which to take input data
+    <model_dir>: where the pretrained model is located
+    <cyc_len>: cycle length (also the number of epochs -- only 1 cycle is done)
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: pls explain better

File: ulmfit-fastai-v1/finetune_lm.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    <lr>: learning rate at the last layer
+    <lr_factor>: learning rate of layer n is layerning rate at layer (n+1)
+        times this
+    """
+    data_lm = load_data(data_dir, "data_lm.pkl")
+    
+    awd_lstm_lm_config = dict(
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: load hparams from json, as it was added to main.py

File: ulmfit-fastai-v1/finetune_lm.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        emb_sz=100, n_hid=256, n_layers=3, pad_token=1, qrnn=False, bidir=False, output_p=0.1,
+        hidden_p=0.15, input_p=0.25, embed_p=0.02, weight_p=0.2, tie_weights=True, out_bias=True
+    )
+    learner = lm_learner(data_lm, AWD_LSTM, model_dir, pretrained=True, config=awd_lstm_lm_config)
+    learner.model_dir = ""
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: add new parameter - dir for the finetuned model and save the encoder there instead of using this current local path

File: ulmfit-fastai-v1/finetune_lm.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    learner.unfreeze()
+    
+    # Calculate learning rates for each layer.
+    num_layers = len(learner.layer_groups)
+    lrs = [lr * lr_factor**i for i in range(num_layers)][::-1]
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: pep8 - spaces around **

File: ulmfit-fastai-v1/finetune_lm.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    
+    learner.fit_one_cycle(cyc_len = cyc_len, max_lr = lrs, div_factor = 32, pct_start = 0.1)
```
Comment (maaario) [inline ulmfit-fastai-v1/finetune_lm.py]: pep8 - no spaces around = in kwargs

File: ulmfit-fastai-v1/train_clas.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
+import sys, fire
```
Comment (maaario) [inline ulmfit-fastai-v1/train_clas.py]: Tidy up imports as above

File: ulmfit-fastai-v1/train_clas.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+from fastai.text import *
+from utils import *
+
+
+def train_clas(data_dir, model_dir, cyc_len = 1, lr = 4e-3, lr_factor = 1/2.6):
```
Comment (maaario) [inline ulmfit-fastai-v1/train_clas.py]: pep8 - as above

File: ulmfit-fastai-v1/train_clas.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    <model_dir>: where the model (or more concretely, the encoder part)
+        is located; the decoder is constructed ad-hoc
+    <cyc_len>: cycle length (number of epochs for each layer)
+    <lr>: learning rate at the last layer
+    <lr_factor>: learning rate of layer n is layerning rate at layer (n+1)
+        times this
+    """
+    data_lm = load_data(data_dir, "data_lm.pkl")
+    
+    awd_lstm_lm_config = dict(
```
Comment (maaario) [inline ulmfit-fastai-v1/train_clas.py]: load config from json

File: ulmfit-fastai-v1/train_clas.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+        emb_sz=100, n_hid=256, n_layers=3, pad_token=1, qrnn=False, bidir=False, output_p=0.1,
+        hidden_p=0.15, input_p=0.25, embed_p=0.02, weight_p=0.2, tie_weights=True, out_bias=True
+    )
+    learner = lm_learner(data_lm, AWD_LSTM, model_dir, pretrained=False, config=awd_lstm_lm_config)
+    learner.model_dir = ""
```
Comment (maaario) [inline ulmfit-fastai-v1/train_clas.py]: provide path for classifier model as parameter

File: ulmfit-fastai-v1/train_clas.py [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    learner.load_encoder("lm_encoder")
+    
+    # Calculate learning rates for each layer.
+    num_layers = len(learner.layer_groups)
+    lrs = [lr * lr_factor**i for i in range(num_layers)][::-1]
```
Comment (maaario) [inline ulmfit-fastai-v1/train_clas.py]: pep8 - no spaces around **

Comment (buj) [reply ulmfit-fastai-v1/train_clas.py]: PEP8 actually has no problem with this, I also believe it does not hinder readability.

File: ulmfit-fastai-v1/train_clas.py (hunk 16 above)
Comment (buj) [reply ulmfit-fastai-v1/train_clas.py]: done

File: ulmfit-fastai-v1/train_clas.py (hunk 15 above)
Comment (buj) [reply ulmfit-fastai-v1/train_clas.py]: done

File: ulmfit-fastai-v1/train_clas.py (hunk 14 above)
Comment (buj) [reply ulmfit-fastai-v1/train_clas.py]: done

File: ulmfit-fastai-v1/finetune_lm.py (hunk 9 above)
Comment (buj) [reply ulmfit-fastai-v1/finetune_lm.py]: should be more in-detail now

File: ulmfit-fastai-v1/finetune_lm.py (hunk 8 above)
Comment (buj) [reply ulmfit-fastai-v1/finetune_lm.py]: removed

File: ulmfit-fastai-v1/finetune_lm.py (hunk 7 above)
Comment (buj) [reply ulmfit-fastai-v1/finetune_lm.py]: spaces around "=" removed, spaces around "/" are not mandated by PEP8.

File: ulmfit-fastai-v1/finetune_lm.py (hunk 6 above)
Comment (buj) [reply ulmfit-fastai-v1/finetune_lm.py]: done

File: ulmfit-fastai-v1/finetune_lm.py (hunk 5 above)
Comment (buj) [reply ulmfit-fastai-v1/finetune_lm.py]: done

File: ulmfit-fastai-v1/train_clas.py (hunk 17 above)
Comment (buj) [reply ulmfit-fastai-v1/train_clas.py]: done

File: ulmfit-fastai-v1/README.md [hunk 19]
```diff
@@ -0,0 +1,17 @@
+## Usage
+* lm_dataset.csv - csv file, where in the first column are texts
+* clas_dataset.csv - csv file, where there are texts in the first column and labels in the second
+
```
Comment (adam-dej) [inline ulmfit-fastai-v1/README.md]: It would be nice to also explicitly describe workflows for scenarios in which I don't want to use pre-trained model when fine-tuning, or I want to train model from scratch during classifier training. Legacy pipeline has these options, and it would help us with our goal - determining the added value of fine-tuning a pre-trained model.

**Review by adam-dej — CHANGES_REQUESTED** [review]
Generally I like this re-write. It would be nice to extend the README to include expected formats of inputs, and to describe the structure of files that are being saved to model folders (file names, types, purposes, how to load them outside of our pipeline if needed, …)

I'm sorry that most of my comments are coding style nitpicks. Some of those comments apply to multiple files, but I've written them only once (such as hardcoded tokenizer language code or `os.path` vs Pathlib usage). Please take care to fix all occurences.

File: ulmfit-fastai-v1/train_lm.py [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+def train_lm(data_dir, model_dir, epochs=12, lr=3e-4, pretrained=False, hparam_updates=dict()):
+    """
+    Trains a new language model using the provided dataset.
+
+    Attributes:
+        data_dir (str): A directory with processed training and validation data.
+        model_dir (str): A directory where to store the trained model.
+        epochs (int): Number of epochs for model training.
+        lr (float): Learning rate.
+        pretrained (bool): If `pretrained` is set, a trained model is first loaded from `model_dir`
```
Comment (adam-dej) [inline ulmfit-fastai-v1/train_lm.py]: I believe that renaming this param to `use_pretrained_lm` would help avoid confusion.

File: ulmfit-fastai-v1/train_lm.py [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+            and then it is trained with the provided dataset.
+        hparam_updates (dict): A dictionary with updates of model hyper-parametrs. By default,
```
Comment (adam-dej) [inline ulmfit-fastai-v1/train_lm.py]: A link to fastai docs explaining available options in this dict would be helpful.

**Review by adam-dej — CHANGES_REQUESTED** [review]

File: ulmfit-fastai-v1/train_lm.py (hunk 20 above)
Comment (maaario) [reply ulmfit-fastai-v1/train_lm.py]: ok, I renamed it

File: ulmfit-fastai-v1/train_lm.py (hunk 21 above)
Comment (maaario) [reply ulmfit-fastai-v1/train_lm.py]: a link to code with model is now provided

**Review by adam-dej — APPROVED** [review]
